pkg/globalnet/controllers/global_ingressip_controller.go

```diff
@@ -207,6 +218,40 @@ func (c *globalIngressIPController) onCreate(ingressIP *submarinerv1.GlobalIngre
 			}
 		}
 
+		if err != nil {
+			_ = c.pool.Release(ips...)
+			ingressIP.Status.Conditions = util.TryAppendCondition(ingressIP.Status.Conditions, metav1.Condition{
+				Type:    string(submarinerv1.GlobalEgressIPAllocated),
+				Status:  metav1.ConditionFalse,
+				Reason:  "ProgramIPTableRulesFailed",
+				Message: err.Error(),
+			})
+
+			return true
+		}
+	} else if ingressIP.Spec.Target == submarinerv1.HeadlessServiceEndpoints {
+		endpointsIP := ingressIP.GetAnnotations()[headlessSvcEndpointsIP]
+		if endpointsIP == "" {
+			_ = c.pool.Release(ips...)
+
+			klog.Warningf("%q annotation is missing on %q", headlessSvcEndpointsIP, key)
+
+			return true
+		}
+
+		// TODO: fixme (Just rename and modifying messages?)
+		err = c.iptIface.AddIngressRulesForHeadlessSvcPod(ips[0], endpointsIP)
+		if err != nil {
+			klog.Errorf("Error while programming Service %q ingress rules for Pod: %v", key, err)
+			err = errors.WithMessage(err, "Error programming ingress rules")
+		} else {
+			err = c.iptIface.AddEgressRulesForHeadlessSVCPods(key, endpointsIP, ips[0], globalNetIPTableMark)
+			if err != nil {
+				_ = c.iptIface.RemoveIngressRulesForHeadlessSvcPod(ips[0], endpointsIP)
+				err = errors.WithMessage(err, "Error programming egress rules")
+			}
+		}
+
```

Comment:
This is same as submarinerv1.HeadlessServicePod. We can remove the duplication

Reply:
yeah it differs by annotation key. I suggest to extract a method:

 `func (c *globalIngressIPController) addRulesForHeadlessService(ingressIP *submarinerv1.GlobalIngressIP, annotationKey string, ips []string) bool`

```
if addRulesForHeadlessService(ingressIP, headlessSvcEndpointsIP, ips) {
    return true
}
```

Reply:
Fixed in a bit different way (to be aligned to the other place). `targetType` is passed to add to different chains for pod ("SM-GN-EGRESS-HDLS-PODS") and endpoints ("SM-GN-EGRESS-HDLS-EPS") to keep existing behavior (and it helps  distinguishing the logs).